services/appflowy-collaborate/src/ws2/actors/session.rs

```diff
@@ -0,0 +1,294 @@
+          message,
+          workspace_id: self.current_workspace,
+          object_id,
+          client_id: self.id(),
+          sender: self.info.collab_origin(),
+        });
+      },
+      Err(err) => {
+        tracing::warn!("client {} send invalid message: {}", self.id(), err);
+        ctx.close(Some(CloseReason::from((
+          CloseCode::Invalid,
+          err.to_string(),
+        ))));
+      },
+    }
+  }
+}
+
+impl Actor for WsSession {
+  type Context = ws::WebsocketContext<Self>;
+
+  fn started(&mut self, ctx: &mut Self::Context) {
+    tracing::trace!("starting session `{}`", self.id());
+    self.hb(ctx);
+    let join = Join {
+      session_id: self.id(),
+      collab_origin: self.info.collab_origin(),
+      addr: ctx.address(),
+      last_message_id: self.info.last_message_id.map(MessageId::into),
+      workspace_id: self.current_workspace,
+    };
+    self
+      .server
+      .send(join)
+      .into_actor(self)
+      .then(|res, act, ctx| {
+        if let Err(err) = res {
+          tracing::warn!("session `{}` can't join: {}", act.id(), err);
+          ctx.stop();
+        }
+        fut::ready(())
+      })
+      .wait(ctx);
+  }
+
+  fn stopping(&mut self, _ctx: &mut Self::Context) -> Running {
+    tracing::trace!("stopping session `{}`", self.id());
+    self.server.do_send(Leave {
+      session_id: self.id(),
+      workspace_id: self.current_workspace,
+    });
+    Running::Stop
+  }
+}
+
+impl Handler<WsOutput> for WsSession {
+  type Result = ();
+
+  fn handle(&mut self, msg: WsOutput, ctx: &mut Self::Context) {
+    tracing::trace!(
+      "sending message through session `{}`: {:#?}",
+      self.id(),
+      msg.message
+    );
+    if let Ok(bytes) = msg.message.into_bytes() {
+      ctx.binary(bytes);
+    };
+  }
+}
+
+impl StreamHandler<Result<ws::Message, ws::ProtocolError>> for WsSession {
+  fn handle(&mut self, item: Result<ws::Message, ProtocolError>, ctx: &mut Self::Context) {
+    match item {
+      Ok(message) => {
+        tracing::trace!("received message: {:?}", message);
+        self.hb = Instant::now();
+        match message {
+          ws::Message::Ping(bytes) => {
+            ctx.pong(&bytes);
+          },
+          ws::Message::Pong(_) => {},
+          ws::Message::Text(_) => ctx.close(Some(CloseReason::from((
+            CloseCode::Unsupported,
+            "only binary content is supported",
+          )))),
+          ws::Message::Binary(bytes) => {
+            self.handle_protocol(bytes, ctx);
+          },
+          ws::Message::Continuation(item) => {
+            match item {
+              Item::FirstText(_) => ctx.close(Some(CloseReason::from((
+                CloseCode::Unsupported,
+                "only binary content is supported",
+              )))),
+              Item::FirstBinary(bytes) => self.buf = Some(bytes.into()),
+              Item::Continue(bytes) => {
+                if let Some(ref mut buf) = self.buf {
```

Comment:
Same question applies to unlimited message sizes: what happens if the buffer size is too large? It could cause an OOM error.

Reply:
NVM. I saw the `frame_size(10 * 1024 * 1024)` limit